### DISCUSSION TITLE
posts/postgresql-high-availability/

### DISCUSSION THREAD

**Discussion by giscus** [techno-tim/techno-tim.github.io#334] [Comments]
# posts/postgresql-high-availability/

Today is the day that you start running PostgresSQL in a cluster! In this tutorial we’ll be setting up a production ready Postgres cluster that’s highly available and fault tolerant using PostgreSQL, etcd, Patroni, HA Proxy, and keepalived. This resilient combination will ensure that you can always reach your database even when a node in the cluster goes down!

https://technotim.live/posts/postgresql-high-availability/

**Comment by Po3Tato** [comment]
Great walk through. Typo on the VIP "ping 192.168.60.100" instead of "192.168.60.110". Also i am working to figure out how to have the network being handled by Tailscale. Stuck at the VIP part but gaining some progress.

**Reply by mayrink85** [reply]
Also noticed this @Po3Tato and have submitted a PR to fix this typo.

**Comment by m1ckyb** [comment]
In the Verifying Our Postgres Cluster section, the ports are wrong, you have :8080 but should be 8008 :)

But thanks for this guide..

**Reply by Po3Tato** [reply]
Got a question. Getting this error
'''
 mn-postgresql-srv02 etcd[1358]: {"level":"warn","ts":"2024-12-28T23:42:08.875944Z","caller":"embed/config_logging.go:170","msg":"rejected connection on client endpoint","remote-addr":"127.0.0.1:58096","server-name":"","error":"tls: first record does not look like a TLS handshake"}
'''
My certs look fine. Got any tips?

**Reply by mayrink85** [reply]
Hi @m1ckyb I saw the same thing and have submitted a PR to fix this.

**Reply by niwiunig** [reply]
Hey, may i ask what is the problem, i am stuck at this :(

**Reply by gforke97** [reply]
I got stuck on that too, it works if you pass the cert locations to the etcdctl command like this:
etcdctl endpoint health --cacert="/etc/etcd/ssl/ca.crt" --cert=$(ls /etc/etcd/ssl/etcd-*.crt) --key=$(ls /etc/etcd/ssl/etcd-*.key)
etcdctl member list --cacert="/etc/etcd/ssl/ca.crt" --cert=$(ls /etc/etcd/ssl/etcd-*.crt) --key=$(ls /etc/etcd/ssl/etcd-*.key)

**Reply by gforke97** [reply]
`etcdctl endpoint health --cacert="/etc/etcd/ssl/ca.crt" --cert=$(ls /etc/etcd/ssl/etcd-*.crt) --key=$(ls /etc/etcd/ssl/etcd-*.key)`
`etcdctl member list --cacert="/etc/etcd/ssl/ca.crt" --cert=$(ls /etc/etcd/ssl/etcd-*.crt) --key=$(ls /etc/etcd/ssl/etcd-*.key)`
Markdown formmating removed the wildcards above...

**Comment by MrSliff** [comment]
Hey, thanks for the Guide. 

I am stuck at the Patroni Part, seems Patroni is not able to connect to etcd. I get a 404 not found Error on Patroni Logs. Etcd itself is working fine.

Patroni Logs:

Feb 24 00:00:05 pg3node1.example.com patroni[2328]: 2025-02-24 00:00:05,446 ERROR: Failed to get list of machines from https://192.168.25.102:2380/v3: <Unknown error: '404 page not found', code: 2>
Feb 24 00:00:05 pg3node1..example.com patroni[2328]: 2025-02-24 00:00:05,496 ERROR: Failed to get list of machines from https://192.168.25.101:2380/v3: <Unknown error: '404 page not found', code: 2>
Feb 24 00:00:05 pg3node1..example.com patroni[2328]: 2025-02-24 00:00:05,504 ERROR: Failed to get list of machines from https://192.168.25.103:2380/v3: <Unknown error: '404 page not found', code: 2>
Feb 24 00:00:05 pg3node1.example.com patroni[2328]: 2025-02-24 00:00:05,513 ERROR: KVCache.run <Unknown error: '404 page not found', code: 2>

Etcd:

sudo etcdctl \
>   --endpoints=https://192.168.25.101:2379,https://192.168.25.102:2379,https://192.168.25.103:2379 \
>   --cacert=/etc/etcd/ssl/ca.crt \
>   --cert=/etc/etcd/ssl/etcd-node1.crt \
>   --key=/etc/etcd/ssl/etcd-node1.key \
>   endpoint status --write-out=table
+-----------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
|          ENDPOINT           |        ID        | VERSION | DB SIZE | IS LEADER | IS LEARNER | RAFT TERM | RAFT INDEX | RAFT APPLIED INDEX | ERRORS |
+-----------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
| https://192.168.25.101:2379 | 1a331161db35df36 |  3.5.18 |   20 kB |     false |      false |        16 |         53 |                 53 |        |
| https://192.168.25.102:2379 | 8d19e5479f15a22d |  3.5.18 |   20 kB |     false |      false |        16 |         53 |                 53 |        |
| https://192.168.25.103:2379 | b8905488b0c0e243 |  3.5.18 |   20 kB |      true |      false |        16 |         53 |                 53 |        |
+-----------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+

**Reply by KevinKao809** [reply]
May I know what is HAPROXY version you use? I am using 2.0.33 but got some exception.

**Comment by RampedIndent** [comment]
This tutorial is great, but it is missing one critical thing to be production ready....

It doesn't have backups configured

**Comment by Mailstorm-ctrl** [comment]
Can't seem to get this working. I either get an error such as `"error":"tls: first record does not look like a TLS handshake"` when checking etcd status or I'll get `"tls: private key does not match public key"` which is crazy because they do or I'll get the listening messages but when checking endpoint health I will get an unhealthy cluster for some reason.

**Reply by gforke97** [reply]
you need to pass it the SSL Infos (in the tutorial he just does that without explanation if you scroll down...)
`etcdctl endpoint health --cacert="/etc/etcd/ssl/ca.crt" --cert=$(ls /etc/etcd/ssl/etcd-*.crt) --key=$(ls /etc/etcd/ssl/etcd-*.key)`
`etcdctl member list --cacert="/etc/etcd/ssl/ca.crt" --cert=$(ls /etc/etcd/ssl/etcd-*.crt) --key=$(ls /etc/etcd/ssl/etcd-*.key)`

**Reply by gforke97** [reply]
and later for the patroni commands you need to pass it the config file and for commands that send stuff back like switching nodes -k for insecure...
`patronictl -c /etc/patroni/patroni.yml list`

**Reply by Mailstorm-ctrl** [reply]
I see that now. But I was getting errors before because I was looking at etcd documentation to test the endpoint. I ended up remaking the ca and it's working now for some reason

**Comment by Kamoliddinxoji** [comment]
There may be an issue with the HAProxy timeouts in the postgres_backend section. The following values:
```
timeout connect 5s
timeout client 30s
timeout server 30s
```
are too short for many PostgreSQL clients like pgAdmin or application pools that keep idle connections open. This can cause unexpected disconnections after 30 seconds of inactivity.

**Reply by Iraithium** [reply]
I was thinking the same thing.  What did you end up changing your values to?

**Comment by Judg3d** [comment]
Its great except in my experience (very limited btw) etcd cant handle having nodes down. When i bring any node down for whatever reason (i have a 5 node cluster) the whole thing just freezes

**Comment by dulhaver** [comment]
a comment on your pg_hba settings ... 

- md5   -   is water under the bridge since quite some postgres versions
- scram-sha-256   -   should be used as algorithm for password encryption
